File: trs-sdk-java/src/main/java/dev/trs/sdk/HTTPTransport.java
package dev.trs.sdk;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;

final class HTTPTransport {
    private final String baseUrl;
    private final HttpClient httpClient;
    private final Duration timeout;

    HTTPTransport(String baseUrl, Duration timeout) {
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.timeout = timeout;
        this.httpClient = HttpClient.newBuilder().connectTimeout(timeout).build();
    }

    Object get(String path) {
        return send("GET", path, null);
    }

    Object post(String path, Object body) {
        return send("POST", path, body);
    }

    private Object send(String method, String path, Object body) {
        String rawBody = body == null ? "" : Jsons.stringify(body);
        HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(baseUrl + path))
                .timeout(timeout)
                .header("Accept", "application/json");
        if ("POST".equals(method)) {
            builder = builder
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(rawBody));
        } else {
            builder = builder.GET();
        }
        HttpResponse<String> response;
        try {
            response = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        } catch (IOException | InterruptedException ex) {
            Thread.currentThread().interrupt();
            throw new TRSConnectionError(ex.getMessage(), ex);
        }
        Object payload;
        try {
            payload = response.body().isEmpty() ? Map.of() : Jsons.parse(response.body());
        } catch (RuntimeException ex) {
            throw new TRSProtocolError("invalid JSON response from trs-node");
        }

        if (response.statusCode() >= 200 && response.statusCode() < 300) {
            return payload;
        }
        String message = extractErrorMessage(payload, response.statusCode());
        if (response.statusCode() >= 400 && response.statusCode() < 500) {
            throw new TRSValidationError(message, List.of());
        }
        throw new TRSServerError(message);
    }

    private String extractErrorMessage(Object payload, int statusCode) {
        if (payload instanceof Map<?, ?> map) {
            Object detail = map.get("detail");
            if (detail instanceof String s && !s.isBlank()) {
                return s;
            }
            Object error = map.get("error");
            if (error instanceof String s && !s.isBlank()) {
                return s;
            }
        }
        return "http " + statusCode;
    }
}

File: trs-sdk-kotlin/src/main/kotlin/dev/trs/sdk/HttpTransport.kt
package dev.trs.sdk

import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

internal class HttpTransport(baseUrl: String, timeout: Duration) {
    private val baseUrl: String = baseUrl.trimEnd('/')
    private val timeout: Duration = timeout
    private val client: HttpClient = HttpClient.newBuilder().connectTimeout(timeout).build()

    fun get(path: String): Any? = send("GET", path, null)

    fun post(path: String, payload: Any?): Any? = send("POST", path, payload)

    private fun send(method: String, path: String, payload: Any?): Any? {
        val requestBuilder = HttpRequest.newBuilder(URI.create(baseUrl + path))
            .timeout(timeout)
            .header("Accept", "application/json")
        if (method == "POST") {
            requestBuilder
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(Jsons.stringify(payload)))
        } else {
            requestBuilder.GET()
        }
        val response = try {
            client.send(requestBuilder.build(), HttpResponse.BodyHandlers.ofString())
        } catch (ex: Exception) {
            throw TrsConnectionError(ex.message ?: "connection error", ex)
        }
        val body = response.body().ifBlank { "{}" }
        val parsed = try {
            Jsons.parse(body)
        } catch (ex: Exception) {
            throw TrsProtocolError("invalid JSON response from trs-node")
        }
        if (response.statusCode() in 200..299) {
            return parsed
        }
        val message = extractErrorMessage(parsed, response.statusCode())
        if (response.statusCode() in 400..499) {
            throw TrsValidationError(message, emptyList())
        }
        throw TrsServerError(message)
    }

    private fun extractErrorMessage(payload: Any?, statusCode: Int): String {
        if (payload is Map<*, *>) {
            val detail = payload["detail"] as? String
            if (!detail.isNullOrBlank()) {
                return detail
            }
            val error = payload["error"] as? String
            if (!error.isNullOrBlank()) {
                return error
            }
        }
        return "http $statusCode"
    }
}

